examples/simple_counter/shared/tests/typegen.rs

```diff
@@ -2,7 +2,9 @@ use crux_core::typegen::TypeGen;
 use shared::Counter;
 use std::path::PathBuf;
 
-fn main() -> anyhow::Result<()> {
+#[cfg(feature = "typegen")]
+#[test]
+fn typegen() -> anyhow::Result<()> {
```

Comment:
I'm not sure I understand this change - is the foreign code type generation now done in a test? What's the thinking behind this?

Reply:
Yes, exactly. I learned this pattern from [tokio-console](https://github.com/tokio-rs/console/blob/main/console-api/tests/bootstrap.rs).

The motivation for this change in this PR was to address the following issue as mentioned in description:

> - [x] uniffi-bindgen using a library file failed.
>     - `uniffi-bindgen generate --library target/release/libshared.dylib --language swift --out-dir generate/swift/SharedFF`
>         - Error: `cargo metadata returned 2 packages for crate name shared`
>         - Potential reason: conflict because following crate conflict.
>             - "crate_types":["lib","staticlib","cdylib"]
>             - "crate_types":["bin"]
>         - Potential solution: 
>             - ~~[x] Move `uniffi-bindgen` bin out of `shared` package.~~
>                 - ~~Note: but still failed in 0.28.0 version.~~
>             - ~~[ ] Submit a PR to update `uniffi-bindgen` to ignore crate of bin kind.~~
>                 - It seems that this https://github.com/mozilla/uniffi-rs/pull/2175 solved this issue. 🎉Just wait for version update.
>                 - [ ] Update uniffi to next release version.

I thought the issue with "2 packages for crate name shared" is stemmed from  `shared_types` -> `shared`, so I tried this approach, and removed "shared_types" crate and the dependent relationship.

This method seems simpler for type generation than relying on another crate's `build.rs`, IMO. So I retained this change there. If there are any concerns, I can revert to previous method, as this is not the main issue in this PR. We can discuss further in a new thread then.

Reply:
I've used this pattern before but everywhere else I've seen it has a couple of differences from this implementation:

1. They commit the generated files.  This means you don't need to run some unit tests before parts of the repository will ever build
2. They cause the test to fail if the generated files needed an update - this helps prevent the committed version of the generated files from getting out of date.

When you do it that way I've found it works quite well.  I think it'd be quite awkward otherwise.

Reply:
I do like the direction of eliminating the `shared_types` crate if we can, even though I'm not sure why it's necessary in order to eliminat the `udl` files - the uniffi interface and the generated types are quite independent. 

I'm not convinced test runner is the way to do it though. It's not clear to me whether `tokio-console` does the bootstraping in tests to verify the files it needs (so `cargo test` becomes a kind of readiness self-test), or whether it is genuinely just how you set it up, but for the crux type generation, which is a mandatory step in building the core for use from a "foreign" language, this feels really counter intuitive, however...

What we could do to a similar effect is add a `bin/generate_types.rs` to the `shared` create, which does the job that this test does? I think that `bin` target could then [be configured](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#configuring-a-target) to enable the typegen feature...?

Reply:
Yes, it's necessary to eliminate the 'udl' actually. The true cause is related with:
> It seems that this https://github.com/mozilla/uniffi-rs/pull/2175 solved this issue.

Because it simplify the 'shared_types' crate, so I kept the change. As you suggest, it's a better way to add a `bin/generate_types` to do it, I'll update to this direction.

Reply:
@charypar I've migrated the typegen tests into 'bin/generate-types'. Please take a look.